lib/spack/spack/build_systems/go.py

```diff
@@ -0,0 +1,66 @@
+# Copyright 2013-2019 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+
+from llnl.util.filesystem import mkdirp, install
+from spack.directives import depends_on
+from spack.package import PackageBase, run_after
+from spack.util.executable import Executable
+
+
+class GoPackage(PackageBase):
+    """Specialized class for packages that are built using Go.
+
+    This class provides two phases that can be overridden if required:
+
+        1. :py:meth:`~.GoPackage.build`
+        2. :py:meth:`~.GoPackage.install`
+
+    Packages must add the binaries to installed to
+    :py:attr:`~.GoPackage.executables`.
+
+    Additional build flags can be added to
+    :py:attr:`~.GoPackage.build_flags`.
+
+    :py:meth:`~.GoPackage.setup_build_environment` sets
+    ``GO111MODULE=on`` to enable module mode and sets
+    ``GOFLAGS="-mod-vendor"`` to enable vendoring and prevent ``go``
+    from downloading libraries at build time.
+    """
+
+    #: Phases of a Go package
+    phases = ['build', 'install']
+
+    #: This attribute is used in UI queries that need to know the build
+    #: system base class
+    build_system_class = 'GoPackage'
+
+    depends_on('go@1.11:', type=('build'))
+
+    go = Executable('go')
+    executables = []
+    build_args = []
+
+    def setup_build_environment(self, env):
+        # forcibly enable module mode
+        env.set('GO111MODULE', 'on')
+        # forcibly enable vendoring, prevent network access
+        env.set('GOFLAGS', '-mod=vendor')
```

Comment:
I wonder if we should be using something like `env.append_flags` here.

Reply:
Perhaps.  I'm trying to strike a balance between being authoritarian (using this knob to *force* vendor mode) and flexible.  I don't think that projects will need to set things via GOFLAGS, they're more likely to use `build_args`.  I'd leave it as is and fix it when we have  counter example..

Reply:
Feedback/arguments?

Reply:
Looking at this more, I've decided that I don't want to set either of these in a method that the user can override, they're the things that ensure Spack's promise of no downloads at build time.

I can handle the GOFLAGS by emplacing it on the command line when I call build.

Is there a way to get access to the build environment from with my `install` method, so I can be draconian and set GO111MODULE there?